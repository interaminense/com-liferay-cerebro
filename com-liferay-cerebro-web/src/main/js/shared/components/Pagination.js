import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import Constants from 'shared/util/constants';
import getCN from 'classnames';
import {isFunction, noop, times} from 'lodash';

import Dropdown from './Dropdown';
import Icon from './Icon';
import {setUriQueryValue} from 'shared/util/router';

const DEFAULT_PAGE = Constants.pagination.cur;

const PAGE_BUFFER = 2;

class PaginationEllipsisItem extends Component {
	@autobind
	handleChange() {
		const {onChange, page} = this.props;

		onChange(page);
	}

	render() {
		return (
			<Dropdown.Item {...this.otherProps()} onClick={this.handleChange}>
				{this.props.children}
			</Dropdown.Item>
		);
	}
}

PaginationEllipsisItem.PROPS = {
	onChange: Config.func().value(noop),
	page: Config.number()
};

class PaginationEllipsis extends Component {
	render() {
		const {href, items, onChange} = this.props;

		return (
			<Dropdown
				align="topCenter"
				buttonProps={{
					display: 'link'
				}}
				label="..."
				showCaret={false}
				toggleClasses="page-link"
			>
				{items.map(item => (
					<PaginationEllipsisItem
						href={
							onChange
								? null
								: setUriQueryValue(href, 'page', item)
						}
						key={item}
						onChange={onChange}
						page={item}
					>
						{item}
					</PaginationEllipsisItem>
				))}
			</Dropdown>
		);
	}
}

PaginationEllipsis.PROPS = {
	href: Config.string(),
	items: Config.array(),
	onChange: Config.func()
};

class PaginationItem extends Component {
	@autobind
	handleChange() {
		const {onChange, page} = this.props;

		onChange(page);
	}

	render() {
		const {active, children, disabled, href, onChange, page} = this.props;

		const classes = getCN('page-item', {active, disabled});

		return (
			<li class={classes}>
				{
					do {
						if (page >= 0) {
							if (onChange) {
								<button
									class="page-link"
									onClick={this.handleChange}
								>
									{children}
								</button>;
							} else {
								<a class="page-link" href={href}>
									{children}
								</a>;
							}
						} else {
							children;
						}
					}
				}
			</li>
		);
	}
}

PaginationItem.PROPS = {
	active: Config.bool().value(false),
	disabled: Config.bool().value(false),
	href: Config.string(),
	onChange: Config.func().value(noop),
	page: Config.number().required()
};

class Pagination extends Component {
	getPages() {
		const {href, page, total, onChange} = this.props;

		const pages = times(total, i => i + 1);

		const frontBuffer = page - PAGE_BUFFER;

		if (frontBuffer > 1) {
			const start = 1;
			const numOfItems = frontBuffer - 2;

			const removedItems = pages.slice(start, start + numOfItems);

			pages.splice(
				start,
				numOfItems,
				<PaginationEllipsis
					href={href}
					items={removedItems}
					key="paginationEllipsis1"
					onChange={onChange}
				/>
			);
		}

		const backBuffer = page + PAGE_BUFFER;

		if (backBuffer < total) {
			const start = pages.indexOf(backBuffer + 1);
			const numOfItems = total - backBuffer - 1;

			const removedItems = pages.slice(start, start + numOfItems);

			pages.splice(
				start,
				numOfItems,
				<PaginationEllipsis
					href={href}
					items={removedItems}
					key="paginationEllipsis2"
					onChange={onChange}
				/>
			);
		}

		return pages;
	}

	render() {
		const {href, onChange, page, total} = this.props;

		return (
			<ul class="pagination pagination-root">
				<PaginationItem
					disabled={page === 1}
					href={setUriQueryValue(href, 'page', page - 1)}
					onChange={onChange}
					page={page - 1}
				>
					<Icon symbol="angle-left" />
				</PaginationItem>

				{this.getPages().map((item, index) => (
					<PaginationItem
						active={page === item}
						href={setUriQueryValue(href, 'page', item)}
						key={index}
						onChange={onChange}
						page={isFunction(item) ? -1 : item}
					>
						{item}
					</PaginationItem>
				))}

				<PaginationItem
					disabled={page === total}
					href={setUriQueryValue(href, 'page', page + 1)}
					onChange={onChange}
					page={page + 1}
				>
					<Icon symbol="angle-right" />
				</PaginationItem>
			</ul>
		);
	}
}

Pagination.PROPS = {
	href: Config.string(),
	onChange: Config.func(),
	page: Config.number().value(DEFAULT_PAGE),
	total: Config.number()
};

export default Pagination;