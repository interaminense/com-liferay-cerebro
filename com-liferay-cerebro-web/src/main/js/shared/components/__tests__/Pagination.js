import Pagination from '../Pagination';
import {renderWithStore} from 'test/mock-store';

describe('Pagination', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithStore(Pagination, {href: '', page: 1, total: 1});

		expect(component).toMatchSnapshot();
	});

	it('should render with buttons if passed an onChange handler', () => {
		component = renderWithStore(Pagination, {
			onChange: jest.fn(),
			page: 1,
			total: 1
		});

		expect(component).toMatchSnapshot();
	});

	it('should render with a lot of pages', () => {
		component = renderWithStore(Pagination, {
			href: '',
			page: 1,
			total: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a lot of pages with a middle page active', () => {
		component = renderWithStore(Pagination, {
			href: '',
			page: 50,
			total: 100
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a lot of pages with the end page active', () => {
		component = renderWithStore(Pagination, {
			href: '',
			page: 100,
			total: 100
		});

		expect(component).toMatchSnapshot();
	});
});