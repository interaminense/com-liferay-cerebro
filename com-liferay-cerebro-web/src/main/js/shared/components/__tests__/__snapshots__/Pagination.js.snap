// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render 1`] = `
<ul class="pagination pagination-root">
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
      </svg>
    </button>
  </li>
  <li class="page-item active">
    <button class="page-link">1</button>
  </li>
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
      </svg>
    </button>
  </li>
</ul>
`;

exports[`Pagination should render a lot of pages with a middle page active 1`] = `
<ul class="pagination pagination-root">
  <li class="page-item">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
      </svg>
    </button>
  </li>
  <li class="page-item">
    <button class="page-link">1</button>
  </li>
  <li class="page-item">
    <div class="dropdown dropdown-root" id="overlay2">
      <button class="button-root btn btn-link dropdown-toggle page-link" type="button">
        <span class="text-truncate">... </span>
      </button>
    </div>
  </li>
  <li class="page-item">
    <button class="page-link">48</button>
  </li>
  <li class="page-item">
    <button class="page-link">49</button>
  </li>
  <li class="page-item active">
    <button class="page-link">50</button>
  </li>
  <li class="page-item">
    <button class="page-link">51</button>
  </li>
  <li class="page-item">
    <button class="page-link">52</button>
  </li>
  <li class="page-item">
    <div class="dropdown dropdown-root" id="overlay3">
      <button class="button-root btn btn-link dropdown-toggle page-link" type="button">
        <span class="text-truncate">... </span>
      </button>
    </div>
  </li>
  <li class="page-item">
    <button class="page-link">100</button>
  </li>
  <li class="page-item">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
      </svg>
    </button>
  </li>
</ul>
`;

exports[`Pagination should render a lot of pages with the end page active 1`] = `
<ul class="pagination pagination-root">
  <li class="page-item">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
      </svg>
    </button>
  </li>
  <li class="page-item">
    <button class="page-link">1</button>
  </li>
  <li class="page-item">
    <div class="dropdown dropdown-root" id="overlay4">
      <button class="button-root btn btn-link dropdown-toggle page-link" type="button">
        <span class="text-truncate">... </span>
      </button>
    </div>
  </li>
  <li class="page-item">
    <button class="page-link">98</button>
  </li>
  <li class="page-item">
    <button class="page-link">99</button>
  </li>
  <li class="page-item active">
    <button class="page-link">100</button>
  </li>
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
      </svg>
    </button>
  </li>
</ul>
`;

exports[`Pagination should render with a lot of pages 1`] = `
<ul class="pagination pagination-root">
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
      </svg>
    </button>
  </li>
  <li class="page-item active">
    <button class="page-link">1</button>
  </li>
  <li class="page-item">
    <button class="page-link">2</button>
  </li>
  <li class="page-item">
    <button class="page-link">3</button>
  </li>
  <li class="page-item">
    <div class="dropdown dropdown-root" id="overlay1">
      <button class="button-root btn btn-link dropdown-toggle page-link" type="button">
        <span class="text-truncate">... </span>
      </button>
    </div>
  </li>
  <li class="page-item">
    <button class="page-link">100</button>
  </li>
  <li class="page-item">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
      </svg>
    </button>
  </li>
</ul>
`;

exports[`Pagination should render with buttons if passed an onChange handler 1`] = `
<ul class="pagination pagination-root">
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-left">
        <use xlink:href="/images/lexicon/icons.svg#angle-left"></use>
      </svg>
    </button>
  </li>
  <li class="page-item active">
    <button class="page-link">1</button>
  </li>
  <li class="page-item disabled">
    <button class="page-link">
      <svg class="icon-root lexicon-icon lexicon-icon-angle-right">
        <use xlink:href="/images/lexicon/icons.svg#angle-right"></use>
      </svg>
    </button>
  </li>
</ul>
`;
